### ISSUE TITLE
@file tree disappeared, cannot rebuild it from sentinels

### ISSUE THREAD

**Issue by flimao** [leo-editor/leo-editor#133]
Hello,

I believe I might have found a weird bug in Leo.

I'm working on a TeX file through a Leo `@file` node. I saved the `.leo` file and shut down the computer for the night. On the next day, I open the `.leo` file again, and the `@file` tree is completely empty (only the `@file` node itself is present). The log pane shows an error message:

```
reading: D:\programacao\tex\Projetos\Projeto de Graduacao UFRJ\projeto-leo.leo
reading: @file projeto-leo.tex
----- read error. line: 1, file: D:\programacao\tex\Projetos\Projeto de Graduacao UFRJ\projeto-leo.tex
no tnodeList for <VNode 8a450b8 @file projeto-leo.tex>
write the @file node or use the Import Derived File command
restoreDescendantAttributes: can not find VNode (expanded): gnx = limao.20150124152959.52, tref: limao.20150124152959.52
restoreDescendantAttributes: can not find VNode (expanded): gnx = limao.20150124152959.43, tref: limao.20150124152959.43
restoreDescendantAttributes: can not find VNode (expanded): gnx = limao.20150124200613.41, tref: limao.20150124200613.41
restoreDescendantAttributes: can not find VNode (expanded): gnx = limao.20150124152959.54, tref: limao.20150124152959.54
restoreDescendantAttributes: can not find VNode (expanded): gnx = limao.20150124152959.51, tref: limao.20150124152959.51
restoreDescendantAttributes: can not find VNode (expanded): gnx = limao.20150124152959.53, tref: limao.20150124152959.53
restoreDescendantAttributes: can not find VNode (expanded): gnx = limao.20150124152959.63, tref: limao.20150124152959.63
```
The external file still exists, and the sentinels appear intact, so I tried `Read @<file> nodes`. I only got an error message:

```
reading: @file projeto-leo.tex
orphan node in @file projeto-leo.tex
retaining the data from the .leo file
```

I also tried opening the external file directly in Leo:

```
Traceback (most recent call last):
  File "C:\Program Files (x86)\Leo-5.0-final\leo\core\leoCommands.py", line 556, in doCommand
    val = command(event)
  File "C:\Program Files (x86)\Leo-5.0-final\leo\core\leoCommands.py", line 896, in open
    elif c.looksLikeDerivedFile(fileName):
  File "C:\Program Files (x86)\Leo-5.0-final\leo\core\leoCommands.py", line 821, in looksLikeDerivedFile
    s = f.read()
  File "C:\Python\34x64\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 11651: character maps to <undefined>
```

My system is:

```
Leo 5.0-final, build 20150123105824, Fri, Jan 23, 2015 10:58:24 AM
Git repo info: branch = master, commit = ea695c251c16
Python 3.4.1, PyQt version 5.4.0
Windows 8 AMD64 (build 6.2.9200) 
```
Any ideas?

Cheers,

--Limão

**Comment by flimao** [comment]
Oops,

I forgot to include my environment information:

`Leo Log Window`
`Leo 5.0-final, build 20141124101406, Mon Nov 24 10:14:06 CST 2014`
`Not running from a git repo`
`Python 3.4.1, PyQt version 5.4.0`
`Windows 8 AMD64 (build 6.2.9200) `

Cheers!

**Comment by gatesphere** [comment]
As for the why: that I can't tell.  It seems like a fairly serious bug that @edreamleo will probably have to look into.  Please save your .tex file if possible for further investigation.

To recover it, however: make a copy of your .tex file, and load that into Leo as an @edit node.  That should allow you to recover your data.  When you have the data pulled out of that and organized into a tree, change that to a @file node and save.  Hopefully that will produce a properly formatted file...

**Comment by flimao** [comment]
Thanks for the answer, @gatesphere . Luckily, in this case, I remember how the tree was structured, so it'll be less of a pain in the ass to reassemble it. 

I'll be happy to provide both the `.leo` and the external `.tex` file for analysis.

Cheers!

**Comment by flimao** [comment]
As it turns out, it's not a bug in Leo after all, but a case of [PEBCAK](http://www.urbandictionary.com/define.php?term=PEBCAK).

I created a button to compile the `.tex` file into something useful, like a PDF. Before making the call to the compiler, though, I tried to make sure the external file was clean, by calling `c.atFileCommands.write`. That function writes the external file in a different format (used by the previous version of Leo, I assume), which was why Leo couldn't process the sentinels in it. Since Leo stores meta-information for `@file` nodes on the external files, it was unable to rebuild the tree.

Switching to `c.atFileCommands.writeAll` solves the issue.

Sorry for that!

Cheers!

**Comment by edreamleo** [comment]
On Sun, Jan 25, 2015 at 2:58 PM, Limão <notifications@github.com> wrote:

> I created a button to compile the .tex file into something useful, like a
> PDF. Before making the call to the compiler, though, I tried to make sure
> the external file was clean, by calling c.atFileCommands.write. That
> function writes the external file in a different format (used by the
> previous version of Leo, I assume), which was why Leo couldn't process the
> sentinels in it. Since Leo stores meta-information for @file nodes on the
> external files, it was unable to rebuild the tree.
>
> ​​
> Switching to c.atFileCommands.writeAll solves the issue.
>
> ​An interesting failure mode, though I wouldn't bet that your analysis is
correct:

1. Leo does store data in @file nodes, but that is non-essential data, such
as marks and expansion status.  This data is not needed to recreate a
file.  This is an important principle: external files must be self
contained.

2. at.write should write new-style sentinels by default.  In fact, iirc,
all the existing write code only writes new-style sentinels.

In short, I'm not sure exactly why using at.write failed, but initing reads
and writes is extremely complicated. It's possible there was a unicode
encoding issue involved.

Edward

**Comment by flimao** [comment]
@edreamleo, if you want, I can send you the relevant files so you can check them out.

Before this happened, I assumed (without RTFM) that the `@file` structured was contained in both the external file and the Leo file. As such, I was very surprised when the whole `@file` node structure was just gone!

Is there a directive to make Leo save the `@file` nose structure to the Leo file, as a backup should the external file become corrupted? Or should I use a `@shadow` or a `@nosent` node instead?

Cheers!

**Comment by edreamleo** [comment]
On Sun, Jan 25, 2015 at 9:09 PM, Limão <notifications@github.com> wrote:

> @edreamleo <https://github.com/edreamleo>, if you want, I can send you
> the relevant files so you can check them out.
>

​Please do.  edreamleo@gmail.com

Edward
